### PR TITLE
Rsa sign with sha256 - sha512

### PR DESCRIPTION
**Pull request by DarovskikhAndrei**
use Microsoft Enhanced RSA and AES Cryptographic Provider or Microsoft Enhanced RSA and AES Cryptographic Provider (Prototype) for rsa sign

### CODE REVIEW

File: engines/e_capi.c [hunk 1]
```diff
@@ -888,8 +926,12 @@ int capi_rsa_sign(int dtype, const unsigned char *m, unsigned int m_len,
     /* Now cleanup */
 
  err:
+    if (hash)
     CryptDestroyHash(hash);
```
Comment (vszakats) [inline engines/e_capi.c]: I'd be nice to indent this line.

File: engines/e_capi.c [hunk 2]
```diff
@@ -1561,6 +1613,8 @@ void capi_free_key(CAPI_KEY * key)
     CryptReleaseContext(key->hprov, 0);
     if (key->pcert)
         CertFreeCertificateContext(key->pcert);
+    if (key->id)
```
Comment (richsalz) [inline engines/e_capi.c]: NULL check not needed.

File: engines/e_capi.c [hunk 3]
```diff
@@ -1435,6 +1477,7 @@ static CAPI_KEY *capi_get_key(CAPI_CTX * ctx, const TCHAR *contname,
 {
     CAPI_KEY *key;
     DWORD dwFlags = 0;
+    int len;
```
Comment (richsalz) [inline engines/e_capi.c]: blank line after vars

File: engines/e_capi.c [hunk 4]
```diff
@@ -1435,6 +1477,7 @@ static CAPI_KEY *capi_get_key(CAPI_CTX * ctx, const TCHAR *contname,
...
     key = OPENSSL_malloc(sizeof(CAPI_KEY));
```
Comment (richsalz) [inline engines/e_capi.c]: merge this up on line 1478.  use "*key* as sizeof param

File: engines/e_capi.c (hunk 1 above)
Comment (noloader) [inline engines/e_capi.c]: If `CryptAcquireContext` fails on both tries, then `HCRYPTPROV hprov` will be unintialized and eventually released at `CryptReleaseContext`.

Double failure is not uncommon, especially if the container does not exist.

Comment (DarovskikhAndrei) [inline engines/e_capi.c]: if both CryptAcquireContext fails then hprov = capi_key->hprov;
release context only if hprov && hprov != capi_key->hprov.

File: engines/e_capi.c (hunk 2 above)
Comment (DarovskikhAndrei) [inline engines/e_capi.c]: fix it.

File: engines/e_capi.c (hunk 3 above)
Comment (DarovskikhAndrei) [inline engines/e_capi.c]: fix

File: engines/e_capi.c (hunk 4 above)
Comment (DarovskikhAndrei) [inline engines/e_capi.c]: fix

File: engines/e_capi.c [hunk 5]
```diff
@@ -802,12 +828,13 @@ int capi_rsa_sign(int dtype, const unsigned char *m, unsigned int m_len,
...
 {
     ALG_ID alg;
-    HCRYPTHASH hash;
+    HCRYPTHASH hash = 0;
     DWORD slen;
     unsigned int i;
     int ret = -1;
     CAPI_KEY *capi_key;
     CAPI_CTX *ctx;
+    HCRYPTPROV hprov;
```
Comment (vszakats) [inline engines/e_capi.c]: This should be initialized with zero to cover all the bases.

Comment (DarovskikhAndrei) [inline engines/e_capi.c]: Fixed

Comment (vszakats) [inline engines/e_capi.c]: Thank you!

**Comment by DarovskikhAndrei** [comment]
There are comments in the code?

**Comment by DarovskikhAndrei** [comment]
Have you more comments?

**Comment by mattcaswell** [comment]
Ping @richsalz, what is happening with this? Looks like a new feature so post 1.1.0?

**Comment by richsalz** [comment]
Looks like a new feature for 1.0.2 as initially propsed.

**Comment by mattcaswell** [comment]
Ahhh. Good point. 1.0.2 does not accept new features. If this is still wanted then a new PR should be created against the master branch (although as a feature it is too late for inclusion in 1.1.0). Closing this.

**Pull request closed without merging**